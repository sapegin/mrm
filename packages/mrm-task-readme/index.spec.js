jest.mock('fs');
jest.mock('git-username');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));

const fs = jest.requireActual('fs');
const path = require('path');
const { omitBy } = require('lodash');
const { getTaskOptions } = require('mrm');
const vol = require('memfs').vol;
const task = require('./index');

const stringify = o => JSON.stringify(o, null, '  ');

afterEach(() => vol.reset());

it('should add a readme', async () => {
	vol.fromJSON({
		[`${__dirname}/templates/Readme.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Readme.md'))
			.toString(),
		[`${__dirname}/templates/Contributing.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Contributing.md'))
			.toString(),
		'/package.json': stringify({
			name: 'unicorn',
			repository: 'gandalf/unicorn',
		}),
	});

	task(
		await getTaskOptions(task, false, {
			name: 'Gandalf',
			url: 'https://middleearth.com',
		})
	);

	expect(
		omitBy(vol.toJSON(), (v, k) => k.startsWith(__dirname))
	).toMatchSnapshot();
});

it('should add a readme with custom file name', async () => {
	vol.fromJSON({
		[`${__dirname}/templates/Readme.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Readme.md'))
			.toString(),
		[`${__dirname}/templates/Contributing.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Contributing.md'))
			.toString(),
		'/package.json': stringify({
			name: 'unicorn',
			repository: 'gandalf/unicorn',
		}),
	});

	task(
		await getTaskOptions(task, false, {
			name: 'Gandalf',
			url: 'https://middleearth.com',
			license: 'MIT',
			licenseFile: "${('license-' + license).toUpperCase()}.md",
		})
	);

	expect(vol.toJSON()['/Readme.md']).toMatch(
		'[LICENSE-MIT.md](LICENSE-MIT.md)'
	);
});

it('should add a readme without contributing', async () => {
	vol.fromJSON({
		[`${__dirname}/templates/Readme.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Readme.md'))
			.toString(),
		'/package.json': stringify({
			name: 'unicorn',
			repository: 'gandalf/unicorn',
		}),
	});

	task(
		await getTaskOptions(task, false, {
			name: 'Gandalf',
			url: 'https://middleearth.com',
			includeContributing: false,
		})
	);

	expect(
		omitBy(vol.toJSON(), (v, k) => k.startsWith(__dirname))
	).toMatchSnapshot();
});

it('should add a readme with custom contributing', async () => {
	vol.fromJSON({
		[`${__dirname}/templates/Readme.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Readme.md'))
			.toString(),
		[`${__dirname}/templates/Contributing.md`]: fs
			.readFileSync(path.join(__dirname, 'templates/Contributing.md'))
			.toString(),
		'/package.json': stringify({
			name: 'unicorn',
			repository: 'gandalf/unicorn',
		}),
	});

	task(
		await getTaskOptions(task, false, {
			name: 'Gandalf',
			url: 'https://middleearth.com',
			contributingFile: 'CONTRIBUTIONS.md',
		})
	);

	expect(
		omitBy(vol.toJSON(), (v, k) => k.startsWith(__dirname))
	).toMatchSnapshot();
});
