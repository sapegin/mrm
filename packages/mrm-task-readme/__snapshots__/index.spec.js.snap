// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add a readme 1`] = `
Object {
  "/Readme.md": "# unicorn

...

## Changelog

The changelog can be found on the [Releases page](https://github.com/gandalf/unicorn/releases).

## Contributing

Everyone is welcome to contribute. Please take a moment to review the [contributing guidelines](Contributing.md).

## Authors and license

[Gandalf](https://middleearth.com) and [contributors](https://github.com/gandalf/unicorn/graphs/contributors).

MIT License, see the included [License.md](License.md) file.
",
  "/package.json": "{
  \\"name\\": \\"unicorn\\",
  \\"repository\\": \\"gandalf/unicorn\\"
}",
}
`;
